commons/src/main/java/org/infinispan/commons/configuration/attributes/Attribute.java

```diff
@@ -0,0 +1,211 @@
+package org.infinispan.commons.configuration.attributes;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.infinispan.commons.CacheException;
+
+/**
+ * Attribute. This class implements a configuration attribute value holder. A configuration attribute is defined by an {@link AttributeDefinition}.
+ * An attribute contains an optional value (or defers to its AttributeDefinition for the default value). It is possible to determine whether
+ * a value has been changed with respect to its initial value. An Attribute remains modifiable until it is protected. At which point it can only
+ * be modified if its AttributeDefinition allows it to be mutable. Additionally it is possible to register listeners when values change so that
+ * code can react to these changes.
+ *
+ * @author Tristan Tarrant
+ * @since 7.2
+ */
+public final class Attribute<T> implements Cloneable {
+   private final AttributeDefinition<T> definition;
+   private T value;
+   private boolean protect;
+   private boolean modified;
+   private List<AttributeListener<T>> listeners;
+
+   Attribute(AttributeDefinition<T> definition) {
+      this.definition = definition;
+      this.value = definition.getDefaultValue();
+   }
+
+   public String name() {
+      return definition.name();
+   }
+
+   public T get() {
+      return value;
+   }
+
+   public T getInitialValue() {
+      return definition.getDefaultValue();
+   }
+
+   public void validate() {
+      definition.validate(value);
+   }
+
+   public Attribute<T> protect() {
+      if (!protect && definition.isImmutable()) {
+         this.protect = true;
+      }
+      return this;
+   }
+
+   public void set(T value) {
+      if (protect) {
+         throw new IllegalStateException();
+      }
+      Attribute<T> oldValue = this.clone();
+      this.value = value;
+      this.modified = true;
+      this.fireValueChanged(oldValue);
+   }
+
+   public boolean isImmutable() {
+      return definition.isImmutable();
+   }
+
+   public boolean isModified() {
+      return modified;
+   }
+
+   public boolean asBoolean() {
+      return (Boolean) value;
+   }
+
+   public double asDouble() {
+      return (Double) value;
+   }
+
+   public float asFloat() {
+      return (Float) value;
+   }
+
+   public int asInteger() {
+      return (Integer) value;
+   }
+
+   public long asLong() {
+      return (Long) value;
+   }
+
+   public <K> K asObject(Class<K> klass) {
```

Comment:
The only benefit would be if you inlined the call by avoiding an explicit cast, otherwise as you say the K can be resolved from the compiler checking the type that the variable is defined as.